chainindex/indexer.go

```diff
@@ -0,0 +1,450 @@
+		return xerrors.Errorf("prepare %s: %w", "removeTipsetsBeforeHeightStmt", err)
+	}
+
+	si.removeEthHashesOlderThanStmt, err = si.db.Prepare(stmtRemoveEthHashesOlderThan)
+	if err != nil {
+		return xerrors.Errorf("prepare %s: %w", "removeEthHashesOlderThanStmt", err)
+	}
+
+	si.updateTipsetsToRevertedFromHeightStmt, err = si.db.Prepare(stmtUpdateTipsetsToRevertedFromHeight)
+	if err != nil {
+		return xerrors.Errorf("prepare %s: %w", "updateTipsetsToRevertedFromHeightStmt", err)
+	}
+
+	si.updateEventsToRevertedFromHeightStmt, err = si.db.Prepare(stmtUpdateEventsToRevertedFromHeight)
+	if err != nil {
+		return xerrors.Errorf("prepare %s: %w", "updateEventsToRevertedFromHeightStmt", err)
+	}
+
+	si.isTipsetMessageNonEmptyStmt, err = si.db.Prepare(stmtIsTipsetMessageNonEmpty)
+	if err != nil {
+		return xerrors.Errorf("prepare %s: %w", "isTipsetMessageNonEmptyStmt", err)
+	}
+
+	si.getMinNonRevertedHeightStmt, err = si.db.Prepare(stmtGetMinNonRevertedHeight)
+	if err != nil {
+		return xerrors.Errorf("prepare %s: %w", "getMinNonRevertedHeightStmt", err)
+	}
+
+	si.hasNonRevertedTipsetStmt, err = si.db.Prepare(stmtHasNonRevertedTipset)
+	if err != nil {
+		return xerrors.Errorf("prepare %s: %w", "hasNonRevertedTipsetStmt", err)
+	}
+
+	si.updateEventsToNonRevertedStmt, err = si.db.Prepare(stmtUpdateEventsToNonReverted)
+	if err != nil {
+		return xerrors.Errorf("prepare %s: %w", "updateEventsToNonRevertedStmt", err)
+	}
+
+	si.updateEventsToRevertedStmt, err = si.db.Prepare(stmtUpdateEventsToReverted)
+	if err != nil {
+		return xerrors.Errorf("prepare %s: %w", "updateEventsToRevertedStmt", err)
+	}
+
+	si.getMsgIdForMsgCidAndTipsetStmt, err = si.db.Prepare(stmtGetMsgIdForMsgCidAndTipset)
+	if err != nil {
+		return xerrors.Errorf("prepare %s: %w", "getMsgIdForMsgCidAndTipsetStmt", err)
+	}
+
+	si.insertEventStmt, err = si.db.Prepare(stmtInsertEvent)
+	if err != nil {
+		return xerrors.Errorf("prepare %s: %w", "insertEventStmt", err)
+	}
+
+	si.insertEventEntryStmt, err = si.db.Prepare(stmtInsertEventEntry)
+	if err != nil {
+		return xerrors.Errorf("prepare %s: %w", "insertEventEntryStmt", err)
+	}
+
+	return nil
+}
+
+func (si *SqliteIndexer) IndexEthTxHash(ctx context.Context, txHash ethtypes.EthHash, msgCid cid.Cid) error {
+	si.closeLk.RLock()
+	if si.closed {
+		si.closeLk.RUnlock()
+		return ErrClosed
+	}
+	si.closeLk.RUnlock()
+
+	return withTx(ctx, si.db, func(tx *sql.Tx) error {
+		return si.indexEthTxHash(ctx, tx, txHash, msgCid)
+	})
+}
+
+func (si *SqliteIndexer) IndexSignedMessage(ctx context.Context, msg *types.SignedMessage) error {
+	if msg.Signature.Type != crypto.SigTypeDelegated {
+		return nil
+	}
+	si.closeLk.RLock()
+	if si.closed {
+		si.closeLk.RUnlock()
+		return ErrClosed
+	}
+	si.closeLk.RUnlock()
+
+	return withTx(ctx, si.db, func(tx *sql.Tx) error {
+		return si.indexSignedMessage(ctx, tx, msg)
+	})
+}
+
+func (si *SqliteIndexer) indexSignedMessage(ctx context.Context, tx *sql.Tx, msg *types.SignedMessage) error {
```

Comment:
consider the ordering of these functions, they're a little jumbled here - I think I'd recommend putting the pairs together, so `indexEthTxHash` goes up under `IndexEthTxHash` and the `indexTipsetWithParentEvents`, `indexTipset` and `restoreTipsetIfExists` goes up under `Apply`, with `Revert` by itself.

Reply:
This is done.